### PR TITLE
Add indexed convolutions

### PR DESCRIPTION
**Pull request by lantiga**
This PR adds _indexed convolutions_, i.e. a convolution operator over a 1D tensor (BxCxL), where a list of neighbors for _each element_ is provided through a `indices` tensor (LxK), where K is the size of the convolution kernel. Each row of `indices` specifies the indices of the K neighbors of the corresponding element in the input. A -1 is handled like for zero padding.

Note that the neighbors specified in `indices` are not relative, but rather absolute. They have to be specified _for each_ of the L elements of the input. Relative indices may be added in a separate PR.

A use case for this PR is convolution over non-square lattices, such as images on hexagonal lattices coming from Cherenkov telescopes (http://www.isdc.unige.ch/%7Elyard/FirstLight/FirstLight_slowHD.mov).

Example:
```
import torch

# a 1D input of 5 elems
input = torch.randn(1,1,5)

# this specifies the indices of neighbors for
# each elem of the input (a 3 elem kernel here)
# A -1 corresponds to zero-padding 
indices = torch.ones(5,3).type(torch.LongTensor)

weight = torch.randn(1,1,3)
bias = torch.randn(1)

output = torch.nn.functional.indexed_conv(input, indices, weight, bias)
```

### CODE REVIEW

**Comment by lantiga** [comment]
Note: test fails because of #3810

**Comment by lantiga** [comment]
Ok, now I get a failure on cuda9/cudnn7. My machine still has the previous versions... need to investigate.

**Comment by lantiga** [comment]
Failures were related to implementation issues on my end. These have been fixed now.

**Review by soumith — CHANGES_REQUESTED** [review]
- I dont understand why `indices` has to be of size `(L, K)`. From my understanding, `indices` specifies the particular input indices that the kernel is going to convolve with. `L` in `indices` here is nothing related to input size `L` but is the number of times a dot-product is done as part of the convolution. To give more clarity, in a standard Conv1d, number of dot products = output_size. Here, you are specifying the convention that `output_size = L`, which is why `indices` is of size `L, K`, but there's no harm in relaxing this and making `indices` to be `(output_size, K)`.
- As mentioned in your PR description, indices are absolute and not relative. I want to reinforce that relative indices makes no sense (relative to what?)
- Add tests. Tests for correctness against known inputs / grads, and also jacobian tests.

#### Unsure thoughts
- I think the whole function can be implemented as: Tensor advanced indexing + view + conv1d. There are no downsides (because it uses im2col, there are no memory savings in doing the way you are doing). If so, the entire implementation becomes a trivial few lines of ATen or Python.

#### Refactor request

- move to using ATen. I think the code can be vastly simplified if you make this an ATen native function, rather than the clunky THNN

#### Name Changes

- IndexedConv -> ConvIndexed1d
- indexed_conv -> conv_indexed1d

**Comment by lantiga** [comment]
Thank you @soumith !

> I dont understand why indices has to be of size (L, K). From my understanding, indices specifies the particular input indices that the kernel is going to convolve with. L in indices here is nothing related to input size L but is the number of times a dot-product is done as part of the convolution. To give more clarity, in a standard Conv1d, number of dot products = output_size. Here, you are specifying the convention that output_size = L, which is why indices is of size L, K, but there's no harm in relaxing this and making indices to be (output_size, K).

You're absolutely right. I had my `L == output_size` use case in mind when I wrote the code, but of course there's nothing that constrains it to be that way.

> As mentioned in your PR description, indices are absolute and not relative. I want to reinforce that relative indices makes no sense (relative to what?)

Eh eh, ok. The idea was to eventually provide kernel indices as a column K of relative indices wrt the current input element, for those cases when the arrangement of values in the input follows a regular pattern. It's not a big deal to generate the complete `indices` tensor anyway, so no point to complicate things.

> Add tests. Tests for correctness against known inputs / grads, and also jacobian tests.

Will do.

> Unsure thoughts
> I think the whole function can be implemented as: Tensor advanced indexing + view + conv1d. There are no downsides (because it uses im2col, there are no memory savings in doing the way you are doing). If so, the entire implementation becomes a trivial few lines of ATen or Python.

I remember trying and failing, but it might easily have been me :-) I'll give it another shot.

> Refactor request
> move to using ATen. I think the code can be vastly simplified if you make this an ATen native function, rather than the clunky THNN

I'll happily do that.

> Name Changes
> IndexedConv -> ConvIndexed1d
> indexed_conv -> conv_indexed1d

**Comment by lantiga** [comment]
Ok, for posterity: I went back to searching for an advanced indexing solution and I found a 10-liner that does the same thing as this 1.2k line PR...

```
import torch
from torch.autograd import Variable

def indexed_conv(input, weight, bias, indices):

    nbatch = input.shape[0]
    output_width = indices.shape[0]
    out_chans, in_chans, ksize = weight.shape

    padded = indices == -1
    indices[padded] = 0

    mask = torch.FloatTensor([1,0])
    mask = mask[..., padded.t().long()]
    if isinstance(input, Variable):
        mask = Variable(mask)

    col = input[..., indices.t()] * mask
    col = col.view(nbatch, -1, output_width)

    weight_col = weight.view(out_chans, -1)

    return torch.matmul(weight_col, col) + bias


if __name__ == '__main__':

    input = torch.randn(1,2,5)
    weight = torch.randn(1,2,3)
    bias = torch.randn(1)
    indices = (5 * torch.rand(4,3)).long()
    indices[0,0] = -1

    out = indexed_conv(input, weight, bias, indices)

    input = Variable(input, requires_grad=True)
    weight = Variable(weight)
    bias = Variable(bias)

    out = indexed_conv(input, weight, bias, indices)

    out.sum().backward()

    print(input.grad)
```

It's slightly more memory-inefficient due to the way it handles `-1` "padded" values - it allocates a full `mask` tensor - but it's good enough. Plus it already allows for the output size to be different from the input size.

I'm closing this PR, I'll eventually open a new one.
